webapp/components/generic/SearchableInput/SearchableInput.vue

```diff
@@ -0,0 +1,227 @@
+    role="search"
+    :class="{
+      'is-active': isActive,
+      'is-open': isOpen,
+    }"
+  >
+    <div class="field">
+      <div class="control">
+        <ds-button v-if="isActive" icon="close" ghost class="search-clear-btn" @click="clear" />
+        <ds-select
+          type="search"
+          icon="search"
+          v-model="searchValue"
+          :id="id"
+          label-prop="id"
+          :icon-right="isActive ? 'close' : null"
+          :options="options"
+          :loading="loading"
+          :filter="item => item"
+          :no-options-available="emptyText"
+          :auto-reset-search="!searchValue"
+          :placeholder="$t('search.placeholder')"
+          @click.capture.native="isOpen = true"
+          @focus.capture.native="onFocus"
+          @input.native="handleInput"
+          @keyup.enter.native="onEnter"
+          @keyup.delete.native="onDelete"
+          @keyup.esc.native="clear"
+          @blur.capture.native="onBlur"
+          @input.exact="onSelect"
+        >
+          <template #option="{ option }">
+            <span v-if="isFirstOfType(option)" class="search-heading">
+              <search-heading :resource-type="option.__typename" />
+            </span>
+            <span
+              v-if="option.__typename === 'User'"
+              :class="{ 'option-with-heading': isFirstOfType(option), 'flex-span': true }"
+            >
+              <hc-user :user="option" :showPopover="false" />
+            </span>
+            <span
+              v-if="option.__typename === 'Post'"
+              :class="{ 'option-with-heading': isFirstOfType(option), 'flex-span': true }"
+            >
+              <search-post :option="option" />
+            </span>
+          </template>
+        </ds-select>
+      </div>
+    </div>
+  </div>
+</template>
+<script>
+import { isEmpty } from 'lodash'
+import SearchHeading from '~/components/generic/SearchHeading/SearchHeading.vue'
+import SearchPost from '~/components/generic/SearchPost/SearchPost.vue'
+import HcUser from '~/components/User/User.vue'
+
+export default {
+  name: 'SearchableInput',
+  components: {
+    SearchHeading,
+    SearchPost,
+    HcUser,
+  },
+  props: {
+    id: { type: String },
+    loading: { type: Boolean, default: false },
+    options: { type: Array, default: () => [] },
+  },
+  data() {
+    return {
+      isOpen: false,
+      searchValue: '',
+      value: '',
+      unprocessedSearchInput: '',
+      searchProcess: null,
+      previousSearchTerm: '',
+      delay: 300,
+    }
+  },
+  computed: {
+    emptyText() {
+      return this.isActive && !this.pending ? this.$t('search.failed') : this.$t('search.hint')
+    },
+    isActive() {
+      return !isEmpty(this.previousSearchTerm)
+    },
+  },
+  methods: {
+    isFirstOfType(option) {
+      return (
+        this.options.findIndex(o => o === option) ===
+        this.options.findIndex(o => o.__typename === option.__typename)
+      )
+    },
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-12-20T16:41:04Z" title="Friday, December 20th 2019, 5:41:04 pm +01:00">Dec 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -90,16 +90,14 @@ export default {
     isActive() {
       return !isEmpty(this.previousSearchTerm)
     },
-    isFirstOfType() {
-      return function(option) {
-        return (
-          this.options.findIndex(o => o === option) ===
-          this.options.findIndex(o => o.__typename === option.__typename)
-        )
-      }
-    },
   },
   methods: {
+    isFirstOfType(option) {
+      return (
+        this.options.findIndex(o => o === option) ===
+        this.options.findIndex(o => o.__typename === option.__typename)
+      )
+    },
```


why not @roschaefer ?? it's returning true or false and does not modify internal state
I understood this would make it preferable to be a computed property

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-12-20T16:49:01Z" title="Friday, December 20th 2019, 5:49:01 pm +01:00">Dec 20, 2019</time>_

---


The computed property is set on the component. But `isFirstOfType` is calculated for each data entry.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-01-07T11:26:04Z" title="Tuesday, January 7th 2020, 12:26:04 pm +01:00">Jan 7, 2020</time>_

---


While that makes some sense on the face of it @roschaefer, I'm not satisfied with that answer. First, is that your opinion, or are you referencing some documentation that you could share with us so that we don't just have to take your word for it?
I have done some searching myself, and come to this video recommended in the VueJS docs, which doesn't seem to support your idea... https://vueschool.io/lessons/vuejs-computed-properties?friend=vuejs
Again, it stresses, `"When you need to change data, you will use methods. And when you need to change the presentation of existing data, you will use computed properties."`
This is my understanding of when to prefer one over the other, and does not say anything about when it is for an entire component, or for each data entry.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-01-07T11:28:06Z" title="Tuesday, January 7th 2020, 12:28:06 pm +01:00">Jan 7, 2020</time>_

---


Also, I would draw people's attention to this https://vueschool.io/lessons/vuejs-computed-properties?friend=vuejs
where it talks about the benefit of using a computed property over a method, even though in their example, and our use case, they show the same results, is that `"computed properties are cached based on their reactive dependencies. A computed property will only re-evaluate when some of its reactive dependencies have changed. This means as long as message has not changed, multiple access to the reversedMessage computed property will immediately return the previously computed result without having to run the function again."`